### PR TITLE
US102 Bias Form and State Management

### PR DESCRIPTION
**Pull request by Tengelma**
Overview

Bunch of work related to establishing clear patterns for dynamic form items, vue-ish state management, more abstraction, and just getting a working bias form. 

Abstracted validation for input fields (DoubleField.vue, validators.js)
v-model for input component -> form component (DoubleField and Bias Form)
Updating housing components state using v-model (see top.vue and Bias Form)
Bias form implementation

Includes:

    client/src/components/BiasForm.vue - the bias form
    client/src/components/DoubleField.vue - improvements on FormNumberField for abstracted validation & vue-style state management
    client/src/components/top.vue -- a barebones file intended to show how data flow works using v-model
    client/src/utils/validators.js -- contains functions for use in validating user inputs

Developer Checklist:

The code follows the Quality Policy file on Google Drive [Yes]
I have considered proper use of exceptions [Yes]
I have eliminated IDE warnings [Yes]

### CODE REVIEW

File: client/package.json [hunk 1]
```diff
@@ -11,11 +11,13 @@
...
     "core-js": "^3.6.5",
+    "debug": "^4.2.0",
+    "package.json": "^2.0.1",
     "vue": "^2.6.11",
     "vue-material": "^1.0.0-beta-15",
+    "vue-observe-visibility": "^0.4.6",
     "vue-router": "^3.4.7",
-    "debug": "^4.2.0",
-    "vue-observe-visibility": "^0.4.6"
+    "vuelidate": "^0.7.6"
```
Comment (aneuhold) [inline client/package.json:20]: Looked into how this works. Seems like we might need to learn more about it as a group.

File: client/src/components/top.vue [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      handler() {
+        console.log(this.modelData.bias);
+      },
+    },
+  },
+  components: {
+    BiasForm,
+  },
+};
+</script>
```
Comment (aneuhold) [inline client/src/components/top.vue:30]: What is this component for? Also why is it the only component without a capital first letter? 😛

File: client/package.json [hunk 3]
```diff
@@ -11,11 +11,13 @@
   "dependencies": {
     "axios": "^0.21.0",
     "core-js": "^3.6.5",
+    "debug": "^4.2.0",
+    "package.json": "^2.0.1",
```
Comment (aneuhold) [inline client/package.json:15]: This package says that it is deprecated when installing and to use a different one:

Error

File: client/src/components/BiasForm.vue [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  updated() {
+    this.$emit('onChange', this.model);
+  },
+  watch: {
+    'model.bias_model': function updateOptions(val) {
+      this.model.bias_params = null;
+      this.$nextTick(function () {
+        this.model.bias_params = BACKEND_CONSTANTS.Bias_params[val];
+        this.defaults = BACKEND_CONSTANTS.Bias_params[val];
+      });
```
Comment (aneuhold) [inline client/src/components/BiasForm.vue:70]: This is throwing an unexpected unnamed function error. See below:

Error

This can be fixed by making it an arrow function though. Just checked it and it seems to be updating fine after modifying it and saving changes across model changes as well. See below for an example:

```js
this.$nextTick(() => {
  this.model.bias_params = BACKEND_CONSTANTS.Bias_params[val];
  this.defaults = BACKEND_CONSTANTS.Bias_params[val];
});
```

File: client/src/views/Home.vue [hunk 5]
```diff
@@ -36,8 +37,8 @@
 
 <script>
 import Debug from 'debug';
+import Top from '../components/top.vue';
 import HaloModelForm from '../components/HaloModelForm.vue';
-import CosmologyForm from '../components/CosmologyForm.vue';
```
Comment (aneuhold) [inline client/src/views/Home.vue:40]: Why was this removed? Just tried it with it included and it still works fine.

**Review by aneuhold — CHANGES_REQUESTED** [review]

File: client/src/components/BiasForm.vue (hunk 4 above)
Comment (aneuhold) [reply client/src/components/BiasForm.vue:70]: @aschnapp that would probably be helpful 😁

File: client/package.json (hunk 1 above)
Comment (Tengelma) [reply client/package.json:20]: This wasn't supposed to make it into the pull request. Didn't end up using it. Will be removed in update.

File: client/src/components/top.vue (hunk 2 above)
Comment (Tengelma) [reply client/src/components/top.vue:30]: Just a temporary demo to show how state can be handled using v-model. Will be deleted in final version of this pull request.

Lowercase name is a reminder to delete.

File: client/src/views/Home.vue (hunk 5 above)
Comment (Tengelma) [reply client/src/views/Home.vue:40]: Conflict resolution was a little messy! My bad. Good reason to add tests haha

File: client/src/components/BiasForm.vue (hunk 4 above)
Comment (Tengelma) [reply client/src/components/BiasForm.vue:70]: Will resolve this with next push.

File: client/package.json (hunk 3 above)
Comment (Tengelma) [reply client/package.json:15]: Honestly not sure how/when this made it in.

File: client/src/views/Home.vue (hunk 5 above)
Comment (aneuhold) [reply client/src/views/Home.vue:40]: Haha alright cool. No worries 🎸

**Review by aneuhold — APPROVED** [review]
Looks good 🙂

**Review by aneuhold — CHANGES_REQUESTED** [review]
Oh, it looks like the `CosmologyForm` does actually need to be added back in though for the build to pass.

**Review by vperuzzi — APPROVED** [review]

**Review by aneuhold — APPROVED** [review]
